addons/mail/models/mail_channel.py

```diff
@@ -746,6 +746,41 @@ def _channel_message_notifications(self, message, message_format=False):
     #   - when a message is posted on a channel (to the channel, using _notify() method)
     # ------------------------------------------------------------
 
+    def set_message_pin(self, message_id, pinned):
+        """ (Un)pin a message on the channel and send a notification to the
+        members.
+        :param message_id: id of the message to be pinned.
+        :param pinned: whether the message should be pinned or unpinned.
+        """
+        self.ensure_one()
+        message_to_update = self.message_ids.filtered(
```

Comment:
`self.message_ids` will fetch all messages of channel? It probably doesn't scale.

Why not just use `message_id` param?

Reply:
Changed due to previous comment. What about that?

```python
        message_to_update = self.env['mail.message'].search([
            ['id', '=', message_id],
            ['model', '=', 'mail.channel'],
            ['res_id', '=', self.id],
            ['pinned_at', '=' if pinned else '!=', False]
        ], limit=1)
```